### PR TITLE
v5.0.x: hwloc: enable --enable-plugins when appropriate

### PR DESCRIPTION
**Pull request by awlauria**
When running configure on the internal hwloc, enable plugins via
the --enable-plugins option when 1) --disable-dlopen was not
specified and 2) the user didn't add some variant of
--enable-plugins on the configure line already.  The motivation
for this change is to not pull in hwloc dependencies like libcuda
in the general case.

Signed-off-by: Brian Barrett <bbarrett@amazon.com>
(cherry picked from commit a10882f2512ea5e5c768b4c0d655eeddf991f614)

### CODE REVIEW

**Comment by rhc54** [comment]
This really isn't a fix, you know - it might resolve internal hwloc builds, but you still break if someone points you at an external version. Given how OMPI is typically distributed, that feels more like a bandaid than a solution.

**Comment by awlauria** [comment]
I remember we discussed this on a call and came to the conclusion that external builds should be fine. Is that your recollection @bwbarrett @gpaulsen ?

**Comment by rhc54** [comment]
Not what we are observing with PMIx/PRRTE, neither of which have internal hwloc. Still investigating to try and get to the root cause, but anything over hwloc v2.4.x crashes if built with hwloc default configure.

**Review by janjust — APPROVED** [review]

**Comment by bwbarrett** [comment]
What does crashes mean in this context?

What we decided was that we weren't going to worry about external hwloc with regards to cuda.  Packaged versions of PMIx won't be a problem, because either they won't be built with CUDA support, they'll be built with CUDA support and the packages will pick up the cuda dependency, or the packager will split out the CUDA bits into their own package via --enable-plugins.  Either way, we'll end up something we can link against.  For locally-built HWLOC, we decided there was no solution other than users had to install CUDA everywhere or build with --enable-plugins.  But that's not something that Open MPI can handle.

But none of those things should cause crashes.  Perhaps they would cause loader failures at process startup, but not crashes.  Crashes sound like they would be another bug.

**Comment by rhc54** [comment]
Okay, Josh and I have tracked this down. There are two problems with HWLOC:

1. HWLOC uses "putenv" instead of "setenv". Starting in v2.5.0, he mistakenly "putenv" a string constant, which means we segfault whenever we touch it. This breaks both PMIx and PRRTE as we do manipulate envars.
2. PMIx/PRRTE cannot run without HWLOC's plugins enabled due to the CUDA dependency. We fail every time due to unresolved libraries. Since that isn't a typical configuration, nor the default, we can't work with it.

Bottom line: PMIx/PRRTE are going to have to blacklist HWLOC versions 2.5.0 and above until HWLOC begins releasing fixes. Josh has filed a PR to fix the putenv error. I have no idea what Brice intends to do about the other problem, or if there is something we can do to adapt to it.

**Comment by jjhursey** [comment]
FYI: I filed a PR to fix the specific hwloc issue we found: https://github.com/open-mpi/hwloc/pull/514
